gnark-prover/prover/poseidon/constants.go

```diff
@@ -497,4 +498,417 @@ var CONSTANTS_3 = [][]frontend.Variable{
 		hex("0x216f6717bbc7dedb08536a2220843f4e2da5f1daa9ebdefde8a5ea7344798d22"),
 		hex("0x1da55cc900f0d21f4a3e694391918a1b3c23b2ac773c6b3ef88e2e4228325161"),
 	},
-}
\ No newline at end of file
+}
+
+var MDS_4 = [][]frontend.Variable{
+	{
+		hex("0x236d13393ef85cc48a351dd786dd7a1de5e39942296127fd87947223ae5108ad"),
+		hex("0x277686494f7644bbc4a9b194e10724eb967f1dc58718e59e3cedc821b2a7ae19"),
+		hex("0x023db68784e3f0cc0b85618826a9b3505129c16479973b0a84a4529e66b09c62"),
+		hex("0x1d359d245f286c12d50d663bae733f978af08cdbd63017c57b3a75646ff382c1"),
+	},
+	{
+		hex("0x2a75a171563b807db525be259699ab28fe9bc7fb1f70943ff049bc970e841a0c"),
+		hex("0x083abff5e10051f078e2827d092e1ae808b4dd3e15ccc3706f38ce4157b6770e"),
+		hex("0x1a5ad71bbbecd8a97dc49cfdbae303ad24d5c4741eab8b7568a9ff8253a1eb6f"),
+		hex("0x0d745fd00dd167fb86772133640f02ce945004a7bc2c59e8790f725c5d84f0f"),
+	},
+	{
+		hex("0x2070679e798782ef592a52ca9cef820d497ad2eecbaa7e42f366b3e521c4ed42"),
+		hex("0x2e18c8570d20bf5df800739a53da75d906ece318cd224ab6b3a2be979e2d7eab"),
+		hex("0x0fa86f0f27e4d3dd7f3367ce86f684f1f2e4386d3e5b9f38fa283c6aa723b608"),
+		hex("0x03f3e6fab791f16628168e4b14dbaeb657035ee3da6b2ca83f0c2491e0b403eb"),
+	},
+	{
+		hex("0x2f545e578202c9732488540e41f783b68ff0613fd79375f8ba8b3d30958e7677"),
+		hex("0x23810bf82877fc19bff7eefeae3faf4bb8104c32ba4cd701596a15623d01476e"),
+		hex("0x014fcd5eb0be6d5beeafc4944034cf321c068ef930f10be2207ed58d2a34cdd6"),
+		hex("0x00c15fc3a1d5733dd835eae0823e377f8ba4a8b627627cc2bb661c25d20fb52a"),
+	},
+}
+
+var CONSTANTS_4 = [][]frontend.Variable{
+	{
+		hex("0x19b849f69450b06848da1d39bd5e4a4302bb86744edc26238b0878e269ed23e5"),
+		hex("0x265ddfe127dd51bd7239347b758f0a1320eb2cc7450acc1dad47f80c8dcf34d6"),
+		hex("0x199750ec472f1809e0f66a545e1e51624108ac845015c2aa3dfc36bab497d8aa"),
+		hex("0x157ff3fe65ac7208110f06a5f74302b14d743ea25067f0ffd032f787c7f1cdf8"),
+	},
+	{
+		hex("0x2e49c43c4569dd9c5fd35ac45fca33f10b15c590692f8beefe18f4896ac94902"),

```

Comment:
Constants from the second block and ongoing seem to be inconsistent with circomlib.
(Constants from the block above line 532-535 are consistent [with circomlib](https://github.com/iden3/circomlib/blob/master/circuits/poseidon_constants.circom#L171-L174).)

Reply:
It's interesting that constatnts for t=2 & t=3 are also incosistent with circomlib

https://github.com/iden3/circomlib/blob/master/circuits/poseidon_constants.circom#L12-L13

My output for `sage generate_parameters_grain.sage 1 0 254 2 8 56 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001`:

```
'0x09c46e9ec68e9bd4fe1faaba294cba38a71aa177534cdd1b6c7dc0dbd0abd7a7', '0x0c0356530896eec42a97ed937f3135cfc5142b3ae405b8343c1d83ffa604cb81', '0x1e28a1d935698ad1142e51182bb54cf4a00ea5aabd6268bd317ea977cc154a30', '0x27af2d831a9d2748080965db30e298e40e5757c3e008db964cf9e2b12b91251f', 
...
```

Full log: https://gist.github.com/sergeytimoshin/397ca08bd66af47e5e3b505416cda25b


And circomlib used the constant generator according to comment in their code: https://github.com/iden3/circomlib/blob/master/circuits/poseidon_constants.circom#L2

Reply:
I believe[ this comment](https://github.com/iden3/circomlib/blob/master/circuits/poseidon_constants.circom#L2) just refers to the constants in the for t =2

Reply:
hm, weird.
Did you try to test this with reference values from light-poseidon or circomlibjs?

Reply:
Constants in circomlib diverged in this commit, before that they were the same as my outputs:
https://github.com/iden3/circomlib/commit/738f9bcf3fbd23b318a83373bbba927073affae0

Moreover, constants we use in light-poseidon also diverged from circomlibjs master. 
Here is light-poseidon outputs for t2: https://gist.github.com/sergeytimoshin/774121d594920aea6179333a2a86ec74

And here are circomlib poseidon constants before [plonk commit](https://github.com/iden3/circomlib/commit/738f9bcf3fbd23b318a83373bbba927073affae0), they are the same as ours: https://gist.github.com/sergeytimoshin/86e232d927ee0517bc8cb7137311b2cc

Therefore, our gnark prover use the same constants as we use in light-poseidon, so I think it's safe to merge it.

Reply:
I see, yes constants are consistent with https://github.com/iden3/circomlib/blob/a8196b20101be34efe6d7583ba666574c97b9164/circuits/poseidon_constants.circom#L58-L64

Reply:
also light-poseidon outputs for t2 were not useful since we are looking at t4 here and hex to decimal in circomlib also made it a pain while the light-poseidon constants in bytes are not readable.